classes/decision/types/traits/ToPreviousReviewRound.inc.php

```diff
@@ -0,0 +1,173 @@
+<?php
+
+/**
+ * @file classes/decision/types/traits/ToPreviousReviewRound.inc.php
+ *
+ * Copyright (c) 2014-2022 Simon Fraser University
+ * Copyright (c) 2000-2022 John Willinsky
+ * Distributed under the GNU GPL v3. For full terms see the file docs/COPYING.
+ *
+ * @class decision
+ *
+ * @brief Helper functions for decisions taken in review round to return back to previous round
+ */
+
+namespace PKP\decision\types\traits;
+
+use APP\decision\Decision;
+use APP\submission\Submission;
+use Illuminate\Validation\Validator;
+use PKP\context\Context;
+use PKP\db\DAORegistry;
+use PKP\decision\DecisionType;
+use PKP\decision\Steps;
+use PKP\decision\steps\Email;
+use PKP\mail\mailables\DecisionBackToPreviousReviewRoundNotifyAuthor;
+use PKP\mail\mailables\DecisionReviewerUnassignedNotifyReviewer;
+use PKP\security\Role;
+use PKP\submission\reviewAssignment\ReviewAssignmentDAO;
+use PKP\submission\reviewRound\ReviewRound;
+use PKP\user\User;
+
+trait ToPreviousReviewRound
+{
+    use NotifyReviewers;
+    use NotifyAuthors;
+
+    public function getNewStatus(): ?int
+    {
+        return null;
+    }
+
+    public function getNewReviewRoundStatus(): ?int
+    {
+        return null;
+    }
+
+    public function getLabel(?string $locale = null): string
+    {
+        return __('editor.submission.decision.backToPreviousReviewRound', [], $locale);
+    }
+
+    public function getDescription(?string $locale = null): string
+    {
+        return __('editor.submission.decision.backToPreviousReviewRound.description', [], $locale);
+    }
+
+    public function getCompletedLabel(): string
+    {
+        return __('editor.submission.decision.backToPreviousReviewRound.completed');
+    }
+
+    public function getCompletedMessage(Submission $submission): string
+    {
+        return __('editor.submission.decision.backToPreviousReviewRound.completed.description', ['title' => $submission->getLocalizedFullTitle()]);
+    }
+
+    public function validate(array $props, Submission $submission, Context $context, Validator $validator, ?int $reviewRoundId = null)
```

Comment:
Rather than overwriting this method in the trait, I'd prefer to see a separate method that's called by the class that uses the trait. When a trait overrides a parent class's method (`validate()`), there's no way to combine that functionality with another trait. Consider the following:

```php
trait ToPreviousReviewRound
{
    public function validate(...);
}

trait NotifyEditors
{
    public function validate(...);
}

class ExampleDecision
{
    use ToPreviousReviewRound;
    use NotifyEditors;
}
```

The `validate` methods would clash.

Reply:
Same for `runAdditionalActions` and `getSteps`. If inheritance is really what is wanted here, you can use that instead of a trait.

Reply:
That's a good point. I hadn't thought of that. In such cases, what happens with something like `parent::validate()`?

```php
<?php
trait A {
    public function validate() {
        return parent::validate();
    }
}

class B extends C {
    use A {
        A::validate as aValidate;
    }

    public function validate() {
        $errors = parent::validate();
        $moreErrors = $this->aValidate();
        return array_merge(errors, $moreErrors);
    }
}
```

Would `C::validate()` be called twice in such cases?

Reply:
@NateWr yes , `C::validate()` it will be called twice because of `parent::validate()` defined in `A::validate()`. Thats a good point and I didn't consider such case . 

So what is your suggestion in such case ? 

one option is to have separate methods in the `ToPreviousReviewRound` such as 
```php
class BackToPreviousExternalReviewRound extends DecisionType implements DecisionRetractable
{
    use ToPreviousReviewRound;
    
    public function validate(array $props, Submission $submission, Context $context, Validator $validator, ?int $reviewRoundId = null)
    {
        $this->validateBackoutDecisionToPreviousReviewRound($props, $submission, $context, $validator, $reviewRoundId);
    }

    ...
}
```
and in the `ToPreviousReviewRound` trait we can have 
```php
protected function validateBackoutDecisionToPreviousReviewRound(array $props, Submission $submission, Context $context, Validator $validator, ?int $reviewRoundId = null) 
{
    ... 
}
```
and have similar implementation for `runAdditionalActions` and `getSteps` .

However in our case, the example you provided will only occurred when we have multilevel inheritance as core `validate` method is defined as `abstract` method. So that becomes a reality when one implemented Decision class extends another . One way is to avoid such case by marking each of the decision class as `final`. But again `runAdditionalActions` and `getSteps`  methods are not defined as `abstract`.

Or we can leave it as it is now and modify it only when we encounter such scenario but that is like hanging on `if only that happens` . I will follow and implement any of the approaches as 

1. leave it as it is 
2. making all the decision class as `final`
3. modify the traits as mentioned above

you think is the best for future consideration .

Reply:
as there is already such implementation there as in `IsRecommendation`, I am going to keep it as it is now . If we faced any issue in future , then we can refactor all of such occurrence .